File: backend/matchData.py
from datetime import datetime
import json
import os
import pandas as pd

def match_data(data: dict) -> pd.DataFrame:
    players = {}
    rounds = data["roundResults"]

    pd.set_option('display.max_columns', None)      # Show all columns
    pd.set_option('display.max_rows', None)         # Show all rows
    pd.set_option('display.max_colwidth', None)     # Show full column width
    pd.set_option('display.expand_frame_repr', False)  # Don't wrap DataFrame
    
    # Accumulate headshots and total shots for each player
    for round in rounds:
        for player in round["playerStats"]:
            if player["puuid"] not in players:
                players[player["puuid"]] = {
                    # "name": player["puuid"],
                    "total_kills": 0,
                    "total_shots": 0,
                    "hs_count": 0,
                    "hs_rate": 0.0
                }
            # players[player["puuid"]]["hs_count"] += player["damage"]["headshots"]
            currentPlayer = player["puuid"]
            
            # for name in data["players"]:
            #     if name["puuid"] == currentPlayer:
            #         players[currentPlayer]["name"] = name["gameName"]
            #         players[currentPlayer]["tag"] = name["tagLine"]

            players[currentPlayer]["total_kills"] += len(player["kills"])
            currPlayerDmg = player["damage"]
            for dmg in currPlayerDmg:
                players[currentPlayer]["total_shots"] += dmg["legshots"] + dmg["bodyshots"] + dmg["headshots"]
                players[currentPlayer]["hs_count"] += dmg["headshots"]

    # Calculate headshot rate for each player
    for player in players.values():
        if player["total_shots"] > 0:
            player["hs_rate"] = player["hs_count"] / player["total_shots"] * 100
        else:
            player["hs_rate"] = 0.0

    df = pd.DataFrame.from_dict(players, orient="index")

    # df["match_id"] = data["matchInfo"]["matchId"]
    df["match_date"] = convert_date(data["matchInfo"]["gameStartMillis"])

    df = df.reset_index().rename(columns={"index": "player_id"})
    # print(json.dumps(players, indent=4))
    # print(df)
    return df

def convert_date(date_str: str) -> datetime:
    from datetime import datetime, timezone

    timestamp_ms = date_str
    dt = datetime.fromtimestamp(timestamp_ms / 1000, tz=timezone.utc)

    return dt

File: backend-node/src/matchData.js
/**
 * Match data processing - converts Valorant API match data to structured format.
 */

/**
 * Convert Unix timestamp in milliseconds to Date object.
 * @param {number} timestampMs - Unix timestamp in milliseconds
 * @returns {Date} Date object
 */
function convertDate(timestampMs) {
  return new Date(timestampMs);
}

/**
 * Process raw Valorant match data into structured player statistics.
 * @param {object} data - Raw match data from Valorant API
 * @returns {Array<object>} Array of player statistics
 */
function matchData(data) {
  const players = {};
  const rounds = data.roundResults || [];

  // Accumulate headshots and total shots for each player
  for (const round of rounds) {
    for (const player of round.playerStats || []) {
      const puuid = player.puuid;
      
      if (!players[puuid]) {
        players[puuid] = {
          playerId: puuid,
          totalKills: 0,
          totalShots: 0,
          hsCount: 0,
          hsRate: 0.0,
        };
      }

      players[puuid].totalKills += (player.kills || []).length;
      
      const damage = player.damage || [];
      for (const dmg of damage) {
        players[puuid].totalShots += (dmg.legshots || 0) + (dmg.bodyshots || 0) + (dmg.headshots || 0);
        players[puuid].hsCount += dmg.headshots || 0;
      }
    }
  }

  // Calculate headshot rate for each player
  for (const player of Object.values(players)) {
    if (player.totalShots > 0) {
      player.hsRate = (player.hsCount / player.totalShots) * 100;
    } else {
      player.hsRate = 0.0;
    }
  }

  // Add match date to all players
  const matchDate = convertDate(data.matchInfo?.gameStartMillis || Date.now());
  
  return Object.values(players).map(player => ({
    ...player,
    matchDate,
  }));
}

module.exports = {
  matchData,
  convertDate,
};
